[perf] 추천 생성 로직 비동기 스케줄러로 전환

zerone1202 opened the pull request:
## #️⃣ Issue Number<!--- ex) #이슈번호, #이슈번호 --> 

- Closes #317
- 추천 스케줄러의 실행 이력을 DB에 저장하여 배치 실행 상태를 모니터링할 수 있도록 Phase 1 성능 고도화를 구현했습니다.


## 📝 요약(Summary)<!--- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
- 매일 새벽 2시 자동 실행되는 추천 스케줄러 구현
- 배치 실행 이력 DB 저장 (시작/종료 시간, 성공/실패 건수, 실행 시간)
- 배치 이력 조회 API 제공으로 로그 파일 없이도 모니터링 가능
   - `GET /api/v1/recommendations/batch-history`: 배치 이력 목록 조회 (페이징)
   - `GET /api/v1/recommendations/batch-history/latest`: 최근 배치 이력 조회
- 100명씩 페이징 처리하여 메모리 효율성 확보

 ## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 📸스크린샷 (선택)
- 배치 이력 목록 조회 (페이징) 성공 응답
<img width="1408" height="861" alt="스크린샷 2025-11-12 20 17 32" src="https://github.com/user-attachments/assets/ea3ab2f4-868d-40e8-a74f-f02248af0424" />
<img width="1408" height="861" alt="스크린샷 2025-11-12 20 17 38" src="https://github.com/user-attachments/assets/90c505be-108f-493c-89dc-23f3a06bbd6f" />

- 최근 배치 이력 조회 성공 응답
<img width="1408" height="796" alt="스크린샷 2025-11-12 20 18 02" src="https://github.com/user-attachments/assets/40d99482-c839-4908-9a99-39a72ff57a6d" />


## 💬 공유사항 to 리뷰어<!--- 리뷰어가 중점적으로 봐줬으면 좋겠는 부분이 있으면 적어주세요. -->
<!--- 논의해야할 부분이 있다면 적어주세요.-->
<!--- ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

zerone1202 committed:
[feat] recommendation: 배치 실행 이력 Entity 추가

- 배치 시작/종료 시간, 상태, 처리 결과 추적
- 성공/실패 사용자 수 및 추천 생성 건수 기록
- 실행 시간 측정 및 에러 메시지 저장 기능

zerone1202 committed:
[feat] recommendation: 배치 상태 Enum 추가

- RUNNING, SUCCESS, FAILED 상태 정의

zerone1202 committed:
[feat] recommendation: 배치 이력 Repository 추가

- 배치 이력 조회 (페이징, 기간별, 상태별)
- 최근 배치 이력 및 최근 성공 배치 조회

zerone1202 committed:
[feat] user: 활성 사용자 ID 페이징 조회 메서드 추가

- findAllActiveUserIds() 메서드 추가
- 배치 처리 시 메모리 효율성을 위해 ID만 조회

zerone1202 committed:
[feat] recommendation: 배치 이력 관리 Service 추가

- 배치 시작/성공/실패 기록 관리
- 배치 이력 조회 (페이징, 기간별, 상태별)
- 실행 시간 자동 계산 및 로깅

zerone1202 committed:
[feat] recommendation: 추천 스케줄러 구현

- 매일 새벽 2시 자동 실행 (cron: 0 0 2 * * *)
- 사용자 100명씩 페이징 처리로 메모리 효율성 확보
- 배치 실행 이력 자동 기록 (성공/실패, 처리 건수, 실행 시간)

zerone1202 committed:
[feat] user: 활성 사용자 ID 페이징 조회 메서드 추가

- findAllActiveUserIds() 메서드 추가
- 배치 처리 시 메모리 효율성을 위해 ID만 조회

zerone1202 committed:
[feat] recommendation: 배치 이력 조회 DTO 추가

- RecommendationBatchHistoryResponse: 단건 조회 응답

zerone1202 committed:
[feat] recommendation: 배치 이력 조회 DTO 추가

- RecommendationBatchHistoryListResponse: 목록 조회 응답 (페이징)

zerone1202 committed:
[feat] recommendation: 배치 이력 조회 API 추가

- GET /api/v1/recommendations/batch-history: 배치 이력 목록 조회
- GET /api/v1/recommendations/batch-history/latest: 최근 배치 이력 조회
- 배치 성공/실패, 처리 건수, 실행 시간 모니터링 가능

zerone1202 committed:
[feat] recommendation: 배치 이력 관련 에러 응답 코드 추가

- BATCH_HISTORY_NOT_FOUND: 배치 이력 없음 에러

zerone1202 committed:
[feat] recommendation: 배치 이력 관련 성공 응답 코드 추가

- BATCH_HISTORY_GET_OK: 배치 이력 조회 성공

zerone1202 committed:
[feat] .gitignore에 performance 디렉토리 추가

zerone1202 committed:
[feat] infra: Redis 비밀번호 설정 주석 해제

- REDIS_PASSWORD 설정 활성화
- Docker Redis (--requirepass oot) 설정과 일치하도록 수정
- Redisson 인증 연결 정상화 확인

zerone1202 committed:
[feat] recommendation: 문서 개선

- 수동 추천 생성 API 주석 개선 (스케줄러와의 관계 명시)

zerone1202 committed:
Merge branch 'dev' of https://github.com/Princess-Maker-s-OOT/OOT-Outfit-of-Today into perf/recommendation-318

# Conflicts:
#	.gitignore

zerone1202 committed:
Merge branch 'dev' of https://github.com/Princess-Maker-s-OOT/OOT-Outfit-of-Today into perf/recommendation-318
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

cjn404 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/entity/RecommendationBatchHistory.java:
```diff
@@ -0,0 +1,117 @@
+
+    // 배치 시작 시 초기화
+    public static RecommendationBatchHistory createInitial(LocalDateTime startTime) {
+        return RecommendationBatchHistory.builder()
```
문단 띄어쓰기 필요해보입니다~~

zerone1202 commented:
성능개선 및 API 구현을 동시 진행하여 브랜치명과 커밋메시지 타입이 다릅니다. 참고 부탁드립니다.

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/scheduler/RecommendationScheduler.java:
```diff
@@ -0,0 +1,104 @@
+     * cron 표현식: 초 분 시 일 월 요일
+     * 0 0 2 * * * = 매일 2시 0분 0초
+     */
+    @Scheduled(cron = "0 0 2 * * *")
```
적용하겠습니다.

cjn404 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/scheduler/RecommendationScheduler.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    @Scheduled(cron = "0 0 2 * * *")
+    public void generateDailyRecommendations() {
+        log.info("=== 추천 배치 시작 ===");
```
문단 띄어쓰기가 필요해 보입니다~~

cjn404 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/service/batch/RecommendationBatchHistoryService.java:
```diff
@@ -0,0 +1,45 @@
+
+    // 특정 기간의 배치 이력 조회
+    List<RecommendationBatchHistory> getBatchHistoryByPeriod(
+            LocalDateTime startTime,
```
파라미터가 2개라서 굳이 띄어쓰지 않아도 될 것 같아요~~

cjn404 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/service/batch/command/RecommendationBatchHistoryCommandServiceImpl.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    public RecommendationBatchHistory startBatch() {
+        LocalDateTime now = LocalDateTime.now(clock);
```
문단 띄어쓰기가 필요해 보입니다~~

cjn404 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/service/batch/command/RecommendationBatchHistoryCommandServiceImpl.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    public void completeBatchFailure(Long batchHistoryId, String errorMessage) {
+        RecommendationBatchHistory batchHistory = batchHistoryRepository.findById(batchHistoryId)
```
문단 띄어쓰기가 필요해 보입니다~~

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/controller/RecommendationController.java:
```diff
@@ -158,4 +164,67 @@ public ResponseEntity<Response<SalePostCreateResponse>> createSalePostFromRecomm
+            @RequestParam(defaultValue = "0") int page,
+            @RequestParam(defaultValue = "20") int size
+    ) {
+        Pageable pageable = PageRequest.of(page, size);
```
적용하겠습니다.

cjn404 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/status/BatchStatus.java:
```diff
@@ -0,0 +1,7 @@
+package org.example.ootoutfitoftoday.domain.recommendation.status;
+
+public enum BatchStatus {
+    RUNNING,   // 실행 중
```
문단 띄어쓰기가 필요해 보입니다~~

cjn404 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/service/batch/RecommendationBatchHistoryServiceImpl.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    @Override
+    @Transactional(readOnly = true)
```
쿼리랑 커맨드로 레이어로 쪼개야 하지 않을까요~~

zerone1202 replied:
반영했습니다.

cjn404 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/repository/RecommendationBatchHistoryRepository.java:
```diff
@@ -0,0 +1,36 @@
+            "WHERE rbh.startTime >= :startTime AND rbh.startTime < :endTime " +
+            "ORDER BY rbh.startTime DESC")
+    List<RecommendationBatchHistory> findByStartTimeBetween(
+            LocalDateTime startTime,
```
파라미터가 2개라서 띄어쓰지 않아도 될 것 같아요~

cjn404 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/controller/RecommendationController.java:
```diff
@@ -158,4 +164,67 @@ public ResponseEntity<Response<SalePostCreateResponse>> createSalePostFromRecomm
+    )
+    @GetMapping("/batch-history/latest")
+    public ResponseEntity<Response<RecommendationBatchHistoryResponse>> getLatestBatchHistory() {
+        RecommendationBatchHistory batchHistory = batchHistoryService.getLastBatchHistory();
```
문단 띄어쓰기가 필요해 보입니다~~

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/scheduler/RecommendationScheduler.java:
```diff
@@ -0,0 +1,104 @@
+        log.info("=== 추천 배치 시작 ===");
+
+        // 배치 이력 시작 기록
+        RecommendationBatchHistory batchHistory = batchHistoryService.startBatch();
```
적용하겠습니다.

zerone1202 commented:
@cjn404  지나님 말씀주신 부분 확인 후 수정해두겠습니다. 개별로 답글 달기 어려워 답글 따로 남깁니다.

cjn404 reviewed (approved):
고생하셧습니다!

AllSungho reviewed (approved):
고생하셨습니다!

zerone1202 committed:
[feat] recommendation: 배치 이력 Command Service 분리

  CQRS 패턴 적용 - Command 영역:
  - startBatch(): 배치 시작 기록
  - completeBatchSuccess(): 배치 성공 기록
  - completeBatchFailure(): 배치 실패 기록
  - handleStaleBatches(): Stale 배치 처리

zerone1202 committed:
[feat] recommendation: 배치 이력 Query Service 분리

- CQRS 패턴 적용 - Query 영역:
 - getRecentBatchHistory(): 최근 배치 이력 조회
 - getBatchHistoryByPeriod(): 기간별 배치 이력 조회
 - getBatchHistoryByStatus(): 상태별 배치 이력 조회
 - getLastBatchHistory(): 최근 배치 이력 단건 조회
 - getLastSuccessBatchHistory(): 최근 성공 배치 이력 조회

zerone1202 committed:
[feat] recommendation: 스케줄러 CQRS 패턴 적용

- RecommendationBatchHistoryService -> RecommendationBatchHistoryCommandService로 변경
- Command 작업만 사용하도록 수정
- 개행 정리

zerone1202 committed:
[feat] recommendation: Controller CQRS 패턴 적용

- RecommendationBatchHistoryService -> RecommendationBatchHistoryQueryService로 변경
- Query 작업만 사용하도록 수정
- 정렬 기준 필드 및 정렬 방향 추가

zerone1202 committed:
[feat] recommendation: 기존 배치 이력 Service 파일 삭제

- CQRS 패턴 적용으로 Command/Query로 분리되어 기존 파일 제거

zerone1202 committed:
[feat] recommendation: 개행 정리

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/scheduler/RecommendationScheduler.java:
```diff
@@ -0,0 +1,104 @@
+        AtomicInteger failedUsers = new AtomicInteger(0);
+        AtomicInteger totalRecommendations = new AtomicInteger(0);
+
+        try {
+            int pageNumber = 0;
+            Page<Long> userIdsPage;
+
+            // 사용자를 100명씩 페이징하여 순차 처리
+            do {
+                userIdsPage = userQueryService.findAllActiveUserIds(PageRequest.of(pageNumber, PAGE_SIZE));
+
+                log.info("Processing page {}: {} users", pageNumber, userIdsPage.getNumberOfElements());
+
+                // 각 사용자에 대해 추천 생성
+                for (Long userId : userIdsPage.getContent()) {
+                    totalUsers.incrementAndGet();
+
+                    try {
+                        int recommendationCount = recommendationCommandService
+                                .generateRecommendations(userId)
+                                .size();
```
추후 고려해보겠습니다.

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/scheduler/RecommendationScheduler.java:
```diff
@@ -0,0 +1,104 @@
+                for (Long userId : userIdsPage.getContent()) {
+                    totalUsers.incrementAndGet();
+
+                    try {
+                        int recommendationCount = recommendationCommandService
+                                .generateRecommendations(userId)
+                                .size();
+
+                        totalRecommendations.addAndGet(recommendationCount);
+                        successUsers.incrementAndGet();
```
이 부분도 추후 고려해보겠습니다.

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/scheduler/RecommendationScheduler.java:
```diff
@@ -0,0 +1,104 @@
+                            log.debug("Generated {} recommendations for user {}", recommendationCount, userId);
+                        }
+
+                    } catch (Exception e) {
+                        failedUsers.incrementAndGet();
+                        log.error("Failed to generate recommendations for user {}: {}", userId, e.getMessage(), e);
+                    }
```
추후 고려해보겠습니다.

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/controller/RecommendationController.java:
```diff
@@ -158,4 +164,67 @@ public ResponseEntity<Response<SalePostCreateResponse>> createSalePostFromRecomm
+    @GetMapping("/batch-history")
+    public ResponseEntity<Response<RecommendationBatchHistoryListResponse>> getBatchHistory(
+            @RequestParam(defaultValue = "0") int page,
+            @RequestParam(defaultValue = "20") int size
```
추후 고려해보겠습니다.

zerone1202 committed:
[feat] recommendation: 실행 시간 계산 로직 중복 제거

- updateExecutionTime() private 헬퍼 메서드 추출
- markAsSuccess()와 markAsFailed()에서 공통으로 사용
- Duration.between() 중복 코드 제거

zerone1202 commented on src/main/java/org/example/ootoutfitoftoday/domain/recommendation/entity/RecommendationBatchHistory.java:
```diff
@@ -0,0 +1,117 @@
+        this.successUsers = successUsers;
+        this.failedUsers = failedUsers;
+        this.totalRecommendations = totalRecommendations;
+        this.executionTimeMs = java.time.Duration.between(startTime, endTime).toMillis();
```
반영했습니다.

AllSungho reviewed (approved):
수고하셨습니다!!
